pkg/providers/subnet/subnet.go

```diff
@@ -69,6 +75,8 @@ func NewDefaultProvider(ec2api ec2iface.EC2API, cache *cache.Cache, availableIPA
 		associatePublicIPAddressCache: associatePublicIPAddressCache,
 		// inflightIPs is used to track IPs from known launched instances
 		inflightIPs: map[string]int64{},
+		// availabilityZoneIDs is used to track possible az IDs based on discovered subnets
+		zoneInfo: map[string]string{},
```

Comment:
We need the zone-id mapping in the cloudprovider `Get` and `List` calls when populating labels on the nodeclaim. I've realized this is actually racey even in our current implementation since we don't know for sure if the EC2NodeClasses have resolved yet. Solving this by introducing a `ZoneProvider`, whose sole responsibility is tracking this mapping. This will have minimal API usage impact since we only need to call this once.

Reply:
Actually realizing it's probably fine to leverage the NodeClass in both cases. By doing this, in the case of upgrade we will fail to provision nodes for pods with zone-id requirements. Once the subnet status is re-reconciled after a minute, we will begin to succeed.